data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/CompressionEngine.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.plugins.source;
+
+import com.amazon.dataprepper.plugins.source.configuration.CompressionOption;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.zip.GZIPInputStream;
+
+public class CompressionEngine {
+    private final CompressionOption compressionOption;
+
+    public CompressionEngine(CompressionOption compressionOption) {
+        this.compressionOption = compressionOption;
+    }
+
+    InputStream createInputStream(String s3Key, InputStream responseInputStream) throws IOException {
+        if (compressionOption.equals(CompressionOption.NONE))
+            return responseInputStream;
+        else if (compressionOption.equals(CompressionOption.GZIP))
+            return new GZIPInputStream(responseInputStream);
+        else {
+            if (s3Key.endsWith(".gz")) {
+                return new GZIPInputStream(responseInputStream);
+            }
+            else {
+                return responseInputStream;
+            }
+        }
```

Comment:
I have a few comments on this code, 1) There is a disconnect with the CompressionOption and the business logic for applying this compressionOption 2) this pattern of leveraging conditionals does not scale well for adding new options. 3) we have to compute the compression engine for every single.

I would recommend we redesign the current class architecture and create an interface for the compression options that requires a `createInputStream`. The desired compression option can be passed in as class parameter to the `S3ObjectWorker`. This will eliminate the redundant computation, provide a tighter coupling for the option and the implementation and make it easier for contributors to add new options.

Reply:
I generally agree with this point. But, I'm willing to defer this change. This is working code that can be applied for 1.5. At this moment we have no plans to add other compression options, nor have I heard of any other formats that might be needed. They might come up in the future (perhaps .zip), but I think we can refactor the code then.

Reply:
Though, if we do it cleanly now, it would be easier for a contributor to add Zip if that contributor needs it.